test/integration/integration_test.go

```diff
@@ -7,54 +7,1073 @@
+
+			claims := defaultClaims()
+			claims["sub"] = "1231"
+			claims["iss"] = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			claims["email"] = "johndoe@example.com"
+
+			signedTokenWithFirstKey, err := idp.Sign(0, claims)
+			Expect(err).NotTo(HaveOccurred())
+
+			makeAPIServerRequestAndExpectCode(signedTokenWithFirstKey, http.StatusForbidden)
+
+			signedTokenWithSecondKey, err := idp.Sign(1, claims)
+			Expect(err).NotTo(HaveOccurred())
+
+			makeAPIServerRequestAndExpectCode(signedTokenWithSecondKey, http.StatusForbidden)
+
+			err = idp.Stop(ctx)
+			Expect(err).NotTo(HaveOccurred())
+
+			waitForOIDCResourceToBeDeleted(ctx, k8sClient, provider)
+		})
+
+		It("Should authenticate but not authorize user with a single registered identity provider and static jwks", func() {
+			idp, err := mockidp.NewIdentityServer("offline", 2)
+			Expect(err).NotTo(HaveOccurred())
+			err = idp.Start()
+			Expect(err).NotTo(HaveOccurred())
+			keys, err := idp.PublicKeySetAsBytes()
+			Expect(err).NotTo(HaveOccurred())
+
+			provider := defaultOIDCProvider()
+			provider.Name = "static"
+			provider.Spec.IssuerURL = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			provider.Spec.JWKS.Keys = keys
+
+			// stop the idp server so that we ensure that the keys will not be fetched over the network
+			err = idp.Stop(ctx)
+			Expect(err).NotTo(HaveOccurred())
+
+			waitForOIDCResourceToBeCreated(ctx, k8sClient, provider)
+			claims := defaultClaims()
+			claims["iss"] = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			claims["sub"] = "my-identity"
+
+			signedTokenWithFirstKey, err := idp.Sign(0, claims)
+			Expect(err).NotTo(HaveOccurred())
+
+			makeAPIServerRequestAndExpectCode(signedTokenWithFirstKey, http.StatusForbidden)
+
+			signedTokenWithSecondKey, err := idp.Sign(1, claims)
+			Expect(err).NotTo(HaveOccurred())
+
+			makeAPIServerRequestAndExpectCode(signedTokenWithSecondKey, http.StatusForbidden)
+
+			waitForOIDCResourceToBeDeleted(ctx, k8sClient, provider)
+		})
+
+		It("Should authenticate and authorize user with a single registered identity provider and defaulted claim to `sub`", func() {
+			idp, err := mockidp.NewIdentityServer("mytestserver", 2)
+			Expect(err).NotTo(HaveOccurred())
+			err = idp.Start()
+			Expect(err).NotTo(HaveOccurred())
+
+			provider := defaultOIDCProvider()
+			provider.Spec.IssuerURL = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+			provider.Spec.CABundle = idp.CA()
+
+			waitForOIDCResourceToBeCreated(ctx, k8sClient, provider)
+
+			user := "this-is-my-identity"
+			claims := defaultClaims()
+			claims["sub"] = user
+			claims["iss"] = fmt.Sprintf("https://localhost:%v", idp.ServerSecurePort)
+
+			signedTokenWithFirstKey, err := idp.Sign(0, claims)
+			Expect(err).NotTo(HaveOccurred())
+
+			roleBinding := createRoleBindingForUser(ctx, defaultNamespaceName, podReaderRoleName, fmt.Sprintf("%s/%s", provider.Name, user))
```

Comment:
```suggestion
			roleBinding := createRoleBindingForUser(ctx, defaultNamespaceName, podReaderRoleName, fmt.Sprintf("%s/%s", provider.Name, user))
			defer deleteRoleBinding(ctx, roleBinding)
```

As the envtest cluster is re-used, you should make sure the role binding is deleted even when some of the assertions fail.